File: cpp/source/backend/ChannelsHandler.cpp
#include "backend/ChannelsHandler.hpp"

#include "backend/ClientBackend.hpp"

using namespace fm;
using namespace fm::backend;

using namespace com::fleetmgr::interfaces;

ChannelsHandler::ChannelsHandler(ClientBackend& _backend) :
    backend(_backend)
{
}

std::shared_ptr<std::vector<traffic::IChannel*>> ChannelsHandler::getChannels()
{
    std::shared_ptr<std::vector<traffic::IChannel*>> result =
            std::make_shared<std::vector<traffic::IChannel*>>();
    result->reserve(channels.size());
    for (auto& pair : channels)
    {
        result->push_back(&(pair.second));
    }
    return result;
}

std::shared_ptr<std::vector<traffic::IChannel*>> ChannelsHandler::getChannels(const std::vector<long>& ids)
{
    std::shared_ptr<std::vector<traffic::IChannel*>> result =
            std::make_shared<std::vector<traffic::IChannel*>>();
    result->reserve(ids.size());
    for (long id : ids)
    {
        result->push_back(&channels.find(id)->second);
    }
    return result;
}

std::shared_ptr<std::vector<traffic::IChannel*>> ChannelsHandler::validateChannels(const std::vector<ChannelResponse>& toValidate)
{
    std::shared_ptr<std::vector<traffic::IChannel*>> result =
            std::make_shared<std::vector<traffic::IChannel*>>();
    for (const ChannelResponse& c : toValidate)
    {
        trace("Opening channel id: " + std::to_string(c.id()));
        std::shared_ptr<traffic::socket::ISocket> socket = backend.createSocket(Protocol::UDP);
        auto pair = channels.emplace(std::piecewise_construct,
                                     std::forward_as_tuple(c.id()),
                                     std::forward_as_tuple(c.id(), socket));
        traffic::ChannelImpl* channel = &pair.first->second;
        if (channel->open(c.host(), c.port(), c.key()))
        {
            trace("Channel id: " + std::to_string(c.id()) + " validated");
            result->push_back(channel);
        }
        else
        {
            channels.erase(c.id());
        }
    }
    return result;
}

void ChannelsHandler::closeChannels(const std::vector<long>& toClose)
{
    for (long id : toClose)
    {
        auto c = channels.find(id);
        if (c != channels.end())
        {
            trace("Closing channel, id: " + std::to_string(id));
            c->second.close();
            channels.erase(c);
        }
        else
        {
            trace("Warning, trying to close not existing channel, id: " + std::to_string(id));
        }
    }
}

void ChannelsHandler::closeAllChannels()
{
    for (auto& pair : channels)
    {
        trace("Closing channel, id: " + std::to_string(pair.first));
        pair.second.close();
    }
    channels.clear();
}

void ChannelsHandler::trace(const std::string& message)
{
    backend.trace(message);
}

File: java/src/main/java/com/fleetmgr/sdk/client/backend/ChannelsHandler.java
package com.fleetmgr.sdk.client.backend;

import com.fleetmgr.interfaces.ChannelIndicationList;
import com.fleetmgr.interfaces.ChannelResponse;
import com.fleetmgr.sdk.client.Client;
import com.fleetmgr.sdk.client.traffic.Channel;
import com.fleetmgr.sdk.client.traffic.ChannelImpl;
import com.fleetmgr.sdk.client.traffic.socket.Socket;
import com.fleetmgr.sdk.client.traffic.socket.UdpSocket;

import java.io.IOException;
import java.util.*;
import java.util.concurrent.ExecutorService;

/**
 * Created by: Bartosz Nawrot
 * Date: 16.12.2018
 * Description:
 */
public class ChannelsHandler {

    private Client client;

    private ExecutorService executor;

    private HashMap<Long, ChannelImpl> channels;

    ChannelsHandler(Client client, ExecutorService executor) {
        this.client = client;
        this.executor = executor;

        this.channels = new HashMap<>();
    }

    public Collection<Channel> getChannels() {
        return new LinkedList<>(channels.values());
    }

    public List<Channel> getChannels(List<Long> channels) {
        LinkedList<Channel> result = new LinkedList<>();
        for (Long id : channels) {
            result.add(this.channels.get(id));
        }
        return result;
    }

    public Map<Long, Channel> validateChannels(Collection<ChannelResponse> toValidate) {
        Map<Long, Channel> opened = new HashMap<>();
        for (ChannelResponse c : toValidate) {
            try {
                trace("Opening channelImpl, id: " + c.getId());

                Socket socket = new UdpSocket(executor);
                ChannelImpl channel = new ChannelImpl(c.getId(), socket);
                channel.open(c.getHost(), c.getPort(), c.getKey());

                channels.put(c.getId(), channel);
                opened.put(c.getId(), channel);

                trace("ChannelImpl id: " + c.getId() + " validated");
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return opened;
    }

    public void closeChannels(Collection<Long> channels) {
        for (Long c : channels) {
            trace("Closing channel, id: " + c);
            ChannelImpl s = this.channels.remove(c);
            if (s != null) {
                s.close();
            } else {
                trace("Warning, trying to close not existing channel, id: " + c);
            }
        }
    }

    public void closeAllChannels() {
        System.out.println(channels);
        for (ChannelImpl c : channels.values()) {
            trace("Closing channel id: " + c.getId());
            c.close();
        }
        System.out.println("asdasdasd");
        channels.clear();
    }

    public void trace(String message) {
        client.trace(message);
    }
}
